Changes to the beforeInteractive strategy to make it work for streaming (#31936)
vercel/next.js: docs/basic-features/script.md

<Script src="https://connect.facebook.net/en_US/sdk.js" strategy="lazyOnload" />
```

There are four different loading strategies that can be used:

- `beforeInteractive`: Load scripts required by the entire site before the page is interactive

> - `beforePageRender`: Load scripts required by a particular page before the page is interactive

Reviewer: Can we separate out `beforePageRender` mode into a separate PR? My understanding was that it would go through the RFC process first, and we need to start showing a warning soon since React 18 is imminent

- `afterInteractive`: (**default**): Load immediately after the page becomes interactive
- `lazyOnload`: Load during idle time

#### beforeInteractive

> Scripts that load with the `beforeInteractive` strategy are injected into the initial HTML from the server and run before self-bundled JavaScript is executed. This strategy should be used for any critical scripts that need to be fetched and executed before the page is interactive. This strategy only works inside **\_document.js** and is designed to load scripts that is needed by the entire site.

Reviewer: Can we add a brief description for why it only works in next/_document?

There are three different loading strategies that can be used:

- `beforeInteractive`: Load script when any page in the application is loaded and before it becomes interactive
<!-- - `beforePageRender`: Load script for a single page before it becomes interactive -->
- `afterInteractive`: (**default**): Load immediately after the page becomes interactive
- `lazyOnload`: Load during idle time

#### beforeInteractive

> Scripts that load with the `beforeInteractive` strategy are injected into the initial HTML from the server and run before self-bundled JavaScript is executed. This strategy should be used for any critical scripts that need to be fetched and executed before any page becomes interactive. This strategy only works inside **\_document.js** and is designed to load scripts that is needed by the entire site (i.e. the script will load when any page in the application has been loaded server-side).

Reviewer: ```suggestion
Scripts that load with the `beforeInteractive` strategy are injected into the initial HTML from the server and run before self-bundled JavaScript is executed. This strategy should be used for any critical scripts that need to be fetched and executed before any page becomes interactive. This strategy only works inside **\_document.js** and is designed to load scripts that are needed by the entire site (i.e. the script will load when any page in the application has been loaded server-side).
```

> The reason `beforeInteractive` was designed this way, to work only inside `\_document.js` is to support streaming modes.

Reviewer: ```suggestion
The reason `beforeInteractive` was designed to work only inside `\_document.js` is to support streaming and Suspense functionality. Outside of the `_document`, it's not possible to guarantee the timing or ordering of `beforeInteractive` scripts.
```

```jsx
// In _document.js

export default class MyDocument extends Document {
  render() {
    return (
      <Html>
        <Head />
        <body>
          <Main />
          <NextScript />
          <Script
            src="https://cdnjs.cloudflare.com/ajax/libs/lodash.js/4.17.20/lodash.min.js"
            strategy="beforeInteractive"
          ></Script>
        </body>
      </Html>
    )
  }
}
```

Examples of scripts that should be loaded as soon as possible with this strategy include:

- Bot detectors
- Cookie consent managers

> <!---

Reviewer: Let's delete this comment for now? (Maybe save on a Git branch)

#### beforeInteractive

Scripts that load with the `beforeInteractive` strategy are injected into the initial HTML from the server and run before self-bundled JavaScript is executed. This strategy should be used for any critical scripts that need to be fetched and executed before any page becomes interactive. This strategy only works inside **\_document.js** and is designed to load scripts that are needed by the entire site (i.e. the script will load when any page in the application has been loaded server-side).

The reason `beforeInteractive` was designed to work only inside `\_document.js` is to support streaming and Suspense functionality. Outside of the `_document`, it's not possible to guarantee the timing or ordering of `beforeInteractive` scripts.

```jsx
// In _document.js

> 

Reviewer 2: ```suggestion
import {Html, Head, Main, NextScript} from 'next/document'
import Script from 'next/script'

```

#### beforeInteractive

Scripts that load with the `beforeInteractive` strategy are injected into the initial HTML from the server and run before self-bundled JavaScript is executed. This strategy should be used for any critical scripts that need to be fetched and executed before any page becomes interactive. This strategy only works inside **\_document.js** and is designed to load scripts that are needed by the entire site (i.e. the script will load when any page in the application has been loaded server-side).

The reason `beforeInteractive` was designed to work only inside `\_document.js` is to support streaming and Suspense functionality. Outside of the `_document`, it's not possible to guarantee the timing or ordering of `beforeInteractive` scripts.

```jsx
// In _document.js

> export default class MyDocument extends Document {

Reviewer 2: Should we use a functional `_document` here as that is what we've been moving docs towards https://nextjs.org/docs/advanced-features/custom-document
